stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -3,78 +3,216 @@
+     * The type member is set to an Eigen Matrix with a scalar type that is the
+     * result of promoting the scalar types of the two provided Eigen Matrix
+     * types
+     *
+     * @tparam T1 Scalar type of first matrix type
+     * @tparam T2 Scalar type of first matrix type
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
      */
     template <typename T1, typename T2, int R, int C>
-    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
-    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
-                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
-      if (!x.empty() && !y.empty()) {
-        check_matching_dims("append_array",
-                            "x[1]",
-                            x.front(),
-                            "y[1]",
-                            y.front());
+    struct append_return_type<Eigen::Matrix<T1, R, C>,
+                              Eigen::Matrix<T2, R, C> > {
+      typedef typename
+      Eigen::Matrix<typename append_return_type<T1, T2>::type, R, C> type;
+    };
+
+    /**
+     * Part of the return type promotion logic for append_array
+     *
+     * If the types of both template arguments are std::vectors, the type member
+     * is recursively computed as the append_return_type of the scalar types
+     * associated with those std::vectors.
+     */
+    template <typename T1, typename T2>
+    struct append_return_type<std::vector<T1>, std::vector<T2> > {
+      typedef typename
+      std::vector<typename append_return_type<T1, T2>::type> type;
+    };
+
+    /**
+     * Part of copy logic for append_array.
+     *
+     * This copies the value at x to y. By default, it is assumed the necessary
+     * copy constructors are available to do the casting from T1 to T2.
+     *
+     * Specializations for std::vector and Eigen::Matrix are provided to do
+     * copy with casting according to the types computed with append_return_type
+     *
+     * @tparam Type of source
+     * @tparam Type of destination
+     * @param x Source variable
+     * @param y Target variable
+     */
+    template<typename T1, typename T2>
+    inline void append_array_copy(const T1& x, T2& y) {
+      y = T2(x);
+    }
+
+    /**
+     * Part of copy logic for append_array.
+     *
+     * The necessary copying/casting for Eigen matrices is delegated to Eigen
+     * template cast
+     *
+     * @tparam Scalar type of source matrix
+     * @tparam Scalar Type of destination matrix
+     * @tparam R Eigen RowsAtCompileTime of both matrices
+     * @tparam C Eigen ColsAtCompileTime of both matrices
+     * @param x Source matrix
+     * @param y Target matrix
+     */
+    template<typename T1, typename T2, int R, int C>
+    inline void append_array_copy(const Eigen::Matrix<T1, R, C>& x,
+                     Eigen::Matrix<T2, R, C>& y) {
+      y = x.template cast<T2>();
```

Comment:
Cool!  I assume that copies all the bits out one by one?  I didn't know that was built in---we do this manually all over the place.